### PR TITLE
DAOS-11408 bio: WAL APIs for replay & checkpoint

### PR DESCRIPTION
**Pull request by NiuYawei**
Implemented repaly & checkpoint APIs, also bio_wal_flush_header()
is exported so that caller could do header flush explicitly.

- bio_meta_readv() & bio_meta_writev()
  These two I/O functions used by caller to load & flush meta pages.

- bio_wal_replay()
  Called on post-crash recovery to replay committed (but not checkpointed)
  transactions into tmpfs. This function assumes the whole meta blob
  has ready being loaded to tmpfs.

- bio_wal_ckp_start()
  Called before starting a checkpoint, caller doesn't have to checkpoint
  to the highest committed tx ID returned by this function.

- bio_wal_ckp_end()
  Called after checkpointing to reclaim WAL space, caller needs to specify
  the highest transaction ID checkpointed to.

- bio_wal_flush_header()
  Explicitly flush WAL header. One use case of bio_wal_flush_header() is
  that vos aggregation can call it to bump the last committed tx ID in WAL
  before each round of aggregation, so that WAL replay on post-crash
  recovery will be able to skip csum verify on the data which might have
  been changed by aggregation.

Required-githooks: true

Signed-off-by: Niu Yawei <yawei.niu@intel.com>

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficent testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by jolivier23 — DISMISSED** [review]

**Review by liuxuezhao — APPROVED** [review]

File: src/bio/bio_wal.c [hunk 1]
```diff
@@ -979,27 +979,628 @@ flush_wal_header(struct bio_meta_context *mc)
...
+{
+	bool	committed = false;
+
+	if (wal_id_cmp(si, tx_id, si->si_commit_id) <= 0)
+		committed = true;
+
+	if (hdr->th_magic != WAL_HDR_MAGIC) {
+		D_CDEBUG(committed, DLOG_ERR, DB_IO, "Mismatched WAL head magic, %x != %x\n",
+			 hdr->th_magic, WAL_HDR_MAGIC);
+		return committed ? -DER_INVAL : 1;
```
Comment (wangshilong) [inline src/bio/bio_wal.c:1046]: what the magic ret '1' means here?

File: src/bio/bio_wal.c [hunk 2]
```diff
@@ -979,27 +979,628 @@ flush_wal_header(struct bio_meta_context *mc)
...
+	csum_len = meta_csum_len(mc);
+	rc = meta_csum_calc(mc, buf, len, &csum, csum_len);
+	if (rc) {
+		D_ERROR("Calculate data csum failed. "DF_RC"\n", DP_RC(rc));
+		return rc;
+	}
+
+	if (csum != expected_csum) {
+		D_DEBUG(DB_IO, "Mismatched data csum, %u != %u\n", csum, expected_csum);
+		return 1;
```
Comment (wangshilong) [inline src/bio/bio_wal.c:1148]: -DER_MISMATCH?

**Review by wangshilong — APPROVED** [review]

File: src/bio/bio_wal.c [hunk 3]
```diff
@@ -979,27 +979,628 @@ flush_wal_header(struct bio_meta_context *mc)
...
+		bio_addr_set(&addr, DAOS_MEDIA_NVME, off2lba(si, off));
+		blks = min(max_blks, nr_blks);
+		if (off + blks > tot_blks)
+			blks = tot_blks - off;
+		bio_iov_set(biov, addr, blks * blk_bytes);
+
+		bsgl.bs_nr_out++;
+		max_blks -= blks;
+		off += blks;
+		if (off == tot_blks)
```
[the remark this answers is not in the document]
Comment (NiuYawei) [reply src/bio/bio_wal.c:1022]: WAL_HDR_BLKS isn't included in 'tot_blks', so as the 'block offset', that's why we need to add header blocks when we try to use SPDK API to read/write blob.

File: src/bio/bio_wal.c (hunk 1 above)
Comment (NiuYawei) [reply src/bio/bio_wal.c:1046]: 1 means replay is done since we hit an uncommitted entry in the WAL.

File: src/bio/bio_wal.c (hunk 2 above)
Comment (NiuYawei) [reply src/bio/bio_wal.c:1148]: It means the data isn't committed (so this transaction isn't committed), and we return 1 to stop replay.

File: src/bio/bio_wal.c [hunk 4]
```diff
@@ -979,27 +979,628 @@ flush_wal_header(struct bio_meta_context *mc)
...
+		goto out;
+	}
+
+	while (1) {
+		hdr = (struct wal_trans_head *)(buf + blk_off * blk_bytes);
+		rc = verify_tx_hdr(si, hdr, tx_id);
+		if (rc)
+			break;
+
+		calc_trans_blks(hdr->th_tot_ents, hdr->th_tot_payload, blk_bytes, &blk_desc);
```
[the remark this answers is not in the document]
Comment (NiuYawei) [reply src/bio/bio_wal.c:1427]: sorry, I'm not sure if I follow your question. Let's discuss it offline.

[the remark this answers is not in the document]
Comment (liuxuezhao) [reply src/bio/bio_wal.c:1427]: no problem, sorry for the noise.

**Comment by NiuYawei** [comment]
The test failures are known OSA extend test failures.

**Pull request merged**